### PR TITLE
RSDK-3319 Add Get Properties to base drivers

### PR DESCRIPTION
**Pull request by randhid**
I'm waiting for RSDK-3301, but I can test with a local version with this API change: https://github.com/viamrobotics/api/compare/main...randhid:api:main. 

- Added `GetProperties` to base and `Properties` wrapper.
- removed LocalBase `Width` method since I don't want two sources of information now that we have WidthM in the `GetProperties` method. It turns out - that's the only thing left in LocalBase, so I removed that too.
- Added injects, tests, and dummies.
- Added driver-specific data for `fake`, `agileX`, and `wheeled`.

@tessavitabile @nfranczak

### CODE REVIEW

File: components/base/wheeled/wheeled_base.go
Comment (edaniels) [inline components/base/wheeled/wheeled_base.go:458]: Isn't this converting it from mm -> micro?

File: components/base/fake/fake.go
Comment (edaniels) [reply components/base/fake/fake.go:86]: yeah I think a component specific Properties in its API makes sense anyway! I cannot think of a good motivation for a common Properties method

File: components/base/fake/fake.go
Comment (randhid) [reply components/base/fake/fake.go:30]: No, just a ballpark possible turning radius for that width based on my experience with steering columns. This is completely changeable.

File: components/base/server_test.go
Comment (randhid) [reply components/base/server_test.go:197]: Yeah, I thought this was weird. This is a pattern in a lot of places elsewhere. Should we follow up to clean up our server/client tests? My team didn't write these tests specifically, but we've been following a pattern set by others.

We can address other components in a follow-up PR, but the client/server tests are kind of very simple.

Comment (edaniels) [reply components/base/server_test.go:197]: yeah we definitely should follow up. They're very subject to copying and pasting so it seems worthwhile. They're low effort and medium value IMO

File: components/base/wheeled/wheeled_base.go
Comment (randhid) [reply components/base/wheeled/wheeled_base.go:457]: no no - the turnign radius is 0.0 if it's not in Ackerman mode. It can spin about it's center.

Comment (edaniels) [reply components/base/wheeled/wheeled_base.go:457]: gotcha. I think when this turns into a struct, we should use the word meters as the suffix. but can we can also leave a comment on the field as it pertains to steering modes?

File: components/base/agilex/limo_base_test.go
Comment (tessavitabile) [inline components/base/agilex/limo_base_test.go:31]: Do you want to add a test for the turning radius as well?

File: components/base/wheeled/wheeled_base.go
Comment (randhid) [reply components/base/wheeled/wheeled_base.go:457]: which is always the case for a wheeled base.

**Review by tessavitabile — COMMENTED** [review]
Thanks for taking this on, @randhid ! Also, if it would unblock merge, I think you could do https://viam.atlassian.net/browse/RSDK-3301.

(No need to wait for my sign-off on the PR.)

File: components/base/client_test.go
Comment (randhid) [reply components/base/client_test.go:45]: ah! forgot to add a test thanks for catching that.

File: components/base/wheeled/wheeled_base.go
Comment (randhid) [reply components/base/wheeled/wheeled_base.go:458]: no that's a division by 0.001, 600mm is .6m 
600mm is 600000um

Comment (edaniels) [reply components/base/wheeled/wheeled_base.go:458]: Damn. Brain not working. Listen, I don't do math, I just program

File: components/base/wheeled/wheeled_base.go
Comment (randhid) [reply components/base/wheeled/wheeled_base.go:457]: sure, can do that.

File: components/base/wheeled/wheeled_base.go
Comment (randhid) [reply components/base/wheeled/wheeled_base.go:458]: It's fine, I'm here to be your opposite.

File: components/base/agilex/limo_base.go
Comment (stevebriskin) [inline components/base/agilex/limo_base.go:442]: Needs mm->m conversion. and in the test.

File: components/base/agilex/limo_base_test.go
Comment (stevebriskin) [reply components/base/agilex/limo_base_test.go:31]: +1. particularly for different steering modes

File: examples/customresources/models/mybase/mybase.go
Comment (stevebriskin) [inline examples/customresources/models/mybase/mybase.go:134]: very minor, not a blocker: i think we should have a dummy implementation (or some comments) here. this example will be used by people to implement new drivers and a nudge in the direction of implementing this method would be good imo. granted, we don't this on other methods but i feel those are more intuitive than this.

File: components/base/agilex/limo_base.go
Comment (stevebriskin) [inline components/base/agilex/limo_base.go:163]: nit: in the future, please have cleanup and variable renaming to separate PRs to make it easier to review the salient changes.

**Review by stevebriskin — APPROVED** [review]
LGTM assuming the unit conversion bug in limo base is taken care of.

File: components/base/agilex/limo_base.go
Comment (randhid) [reply components/base/agilex/limo_base.go:163]: couldn't do that here because I needed to get `base.Features` for get properties, and the created limobase was called `base`.

We shouldn't call any component struct the same as the component package name in the future.

File: components/base/client_test.go
Comment (penguinland) [inline components/base/client_test.go:48]: The old version mentioned the variable `width`, but the new one doesn't seem to use that at all. If that's on purpose, can we remove the `width` argument from `setupWorkingBase`? and if that's not on purpose, please put it back.

File: components/base/client.go
Comment (martha-johnston) [reply components/base/client.go:149]: I think it's because we don't want to be directly using autogenerated code values, or something along those lines

File: components/base/features.go
Comment (randhid) [reply components/base/features.go:14]: Changing this up according to eric's request for this to be astruct rather than a map, let me know if you have more comments after that change.

File: components/base/features.go
Comment (randhid) [reply components/base/features.go:7]: Changing to a structure with types as requested by eric.

File: components/base/client.go
Comment (randhid) [reply components/base/client.go:149]: Yes, we want to wrap the protobuf messages where possible, this is passed in by code generated in the API repo.

File: components/base/features.go
Comment (randhid) [reply components/base/features.go:33]: returning `math.Nan` from a driver is the way to go - different bases will have different error states, and we need the driver being in an error state to signal that to motion somehow. Otherwise, we'd return that and not build the resource in the constructor.

I can also default this to Nan, but would like to talk about what that looks like, would need more extensive tests. Perhaps a follow up ticket.

File: components/base/client_test.go
Comment (penguinland) [reply components/base/client_test.go:48]: Not the `Width` function, the `width` argument (an integer) to the `setupWorkingBase` function.

Comment (randhid) [reply components/base/client_test.go:48]: ah I misunderstood, let me see what that code is doing.

File: components/base/fake/fake.go
Comment (raybjork) [inline components/base/fake/fake.go:88]: [q] We don't want to just store the width as mm?

**Review by raybjork — APPROVED** [review]
Looks a lot cleaner! Love that we're getting rid of LocalBase and the Width method

Will be interested to see how the change to the Feature type ends up looking but don't want to hold off on approving until then

**Review by raybjork — APPROVED** [review]
Looks a lot cleaner! Love that we're getting rid of LocalBase and the Width method

Will be interested to see how the change to the Feature type ends up looking but don't want to hold off on approving until then

File: components/base/fake/fake.go
Comment (randhid) [reply components/base/fake/fake.go:88]: The scope doc for this had meters in the API contract, so changed it here.

Comment (randhid) [reply components/base/fake/fake.go:88]: Math for wheeled base and agileX has it in mm to match `MoveStraight` which expects a MM call, so left that there.

File: components/base/client_test.go
Comment (randhid) [reply components/base/client_test.go:48]: changed it to still take a width, but return that width in Properties as  meters for this test.

File: components/base/client_test.go
Comment (randhid) [reply components/base/client_test.go:183]: changed to distinct error messages

File: examples/customresources/models/mybase/mybase.go
Comment (randhid) [reply examples/customresources/models/mybase/mybase.go]: let me know if you like what I added

File: components/base/agilex/limo_base.go
Comment (randhid) [inline components/base/agilex/limo_base.go:34]: @raybjork, one thing while I was self-reviewing, the scope doc doesn't mention which "width" nav service is going to consume. The wheelbase is what was mentioned in the TP space paper, and `width_mm` is the wheelbase used in the wheeled driver.

The overall width of the base is larger, if it's expected that this parameter be used for obstacles, I can commit these dimensions. I have a feeling for wheeled - no since the `base.CollisionGeometry` takes care of that for built-in motion obstacle avoidance. Let me know if it is in fact the overall width and I'll add a checker in `wheeled` as well for the different frame configs (started there but didn't go too far) 

Made this a separate commit so I can revert it easily.

Comment (tessavitabile) [reply components/base/agilex/limo_base.go:34]: We don't have any plans to use width_mm from GetProperties in motion planning as part of this scope. I think Eliot just thought it should also be exposed via GetProperties. For collisions, motion planning will use the overall width of the base, which it gets from the frame system, I believe.

File: components/base/features.go
Comment (edaniels) [inline components/base/features.go:9]:
```suggestion
type Properties struct {
```

let's match proto

File: components/base/server_test.go
Comment (randhid) [reply components/base/server_test.go]: Made a jira ticket : https://viam.atlassian.net/browse/RSDK-3500

File: components/base/agilex/limo_base.go
Comment (tessavitabile) [reply components/base/agilex/limo_base.go]: width_mm should still be exposed via GetProperties. I'd defer to @raybjork on whether the modules need to have frame configs, since I'm not sure.

Comment (randhid) [reply components/base/agilex/limo_base.go]: yeah width_mm is still exposed, I was reverting the cases where I  expose a different dimension than the wheelbase.

File: components/base/client_test.go
Comment (penguinland) [inline components/base/client_test.go:158]: If each of these sections tests a different thing, should it be a different invocation of `t.Run`? I don't feel strongly either way; keep as-is if that's what you prefer.

File: components/base/fake/fake.go
Comment (penguinland) [inline components/base/fake/fake.go:30]: Maybe I'm being too nitpicky, but why is one of these millimeters and the other meters? Would it make sense to use the same units on both? Maybe not; maybe leave as-is.

File: components/base/properties.go
Comment (penguinland) [inline components/base/properties.go:18]: Typos: truning, minimu

Comment (penguinland) [inline components/base/properties.go:18]: I suspect the comma after "turn around" should be a period.

File: components/base/properties.go
Comment (penguinland) [inline components/base/properties.go:8]: I'm open to disagreement on this, but I think this comment should be shortened. Saying it reports its respective minimum turning radius and width in units of meters does not add anything I couldn't easily get from the code itself (there are two variables, `TurningRadiusMeters` and `WidthMeters`, and it's already clear what they represent).

File: components/base/properties.go
Comment (penguinland) [inline components/base/properties.go:15]: Comment doesn't match code: there is no feature-to-boolean map any more.

File: components/base/properties.go
Comment (penguinland) [inline components/base/properties.go:27]: Comment doesn't match code: there is no longer a map involved.

File: examples/customresources/models/mybase/mybase.go
Comment (penguinland) [inline examples/customresources/models/mybase/mybase.go:140]: I appreciate your consistency, but I think `Truning` should be spelled differently. :smile:

File: components/base/base.go
Comment (penguinland) [inline components/base/base.go:63]: :+1: Thanks for getting rid of `LocalBase` and putting everything into `Base`. I don't see any reason to distinguish local from nonlocal bases.

**Review by penguinland — APPROVED** [review]
My feedback should be easy to address; it's mostly typos in comments. LGTM!

File: components/base/fake/fake.go
Comment (penguinland) [reply components/base/fake/fake.go:30]: Nah, let's keep it as-is for the current PR. You had intentionality behind this, and if we want to clean up the messiness, that can be its own PR later.

File: components/base/properties.go
Comment (randhid) [reply components/base/properties.go:6]: not always, sometimes they're hard-coded into the driver, removing the adjective then.

File: components/base/properties.go
Comment (randhid) [reply components/base/properties.go:15]: Good catches, thanks!

File: components/base/client_test.go
Comment (randhid) [reply components/base/client_test.go:158]: changed, will test with a race and see if I'll commit that

File: components/base/base.go
Comment (randhid) [reply components/base/base.go:63]: Local is just a way to wrap an interface with extra exported functions, ideally we'd only have exported functions that fulfill our API, but sometimes we need that flexibility, in this case, to expose the base's width to someone trying to external math or planning. 

No longer needed though - so got rid of it!

**Pull request merged**
